### PR TITLE
Add onSend hook

### LINKED ISSUE #18: Allow onSend hook to mutate swagger object for each request (multi-tenant hosting)
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

This is a follow-up to [fastify-swagger#683](https://github.com/fastify/fastify-swagger/issues/683).

I have the same need as in https://github.com/fastify/fastify-swagger/pull/92 of setting the authorizationUrl but based not only on the hosting environment but also on the request. The app I have is multi-tenant and each tenant has a different authorizationUrl. Which tenant is calling is determined by the domain.

### Motivation

I don't think this can be achieved with the current postProcessor() introduced in https://github.com/fastify/fastify-swagger/pull/92 since it only runs once at setup.

### Example

For example calling `tenant1.myapp.com/theservice/docs` means that the `authorizationUrl` should be something like `https://login.myapp.com/auth/realms/tenant1/protocol/openid-connect/auth`.

Calling `tenant2.myapp.com/theservice/docs` means that the `authorizationUrl` should be something like `https://login.myapp.com/auth/realms/tenant2/protocol/openid-connect/auth`.

### PR DESCRIPTION
**Pull request by jonaskello**
Fixes #18 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Uzlopak — CHANGES_REQUESTED** [review]
No. It is 2 am now so I will provide arguments tomorrow.

**Comment by jonaskello** [comment]
Maybe I need to update the typescript types manually too or how is that handled?

**Comment by climba03003** [comment]
> Maybe I need to update the typescript types manually too or how is that handled?

Yes, you need to do it manually.

**Comment by jonaskello** [comment]
Ok, added the types. The onSendHook type had a required type param but I made it optional for the uiHooks part so it should be backwards compatible.

Let me know if there is anything else I can do to complete this PR.

File: test/hooks.test.js [hunk 1]
```diff
@@ -74,6 +74,35 @@ test('hooks on static swagger', async t => {
...
 
+test('onSend hook on static swagger', async t => {
+  const fastify = Fastify()
+  await fastify.register(fastifySwagger, {
+    mode: 'static',
+    specification: {
+      path: './examples/example-static-specification.yaml'
+    }
+  })
+  await fastify.register(fastifySwaggerUi, {
```
Comment (Uzlopak) [inline test/hooks.test.js:85]: So you basically adding onSend Hook to all routes. So if you do a request against "/documentation/uiConfig" you get also the modified swagger json.

**Comment by Uzlopak** [comment]
I think this is the wrong approach.

**Comment by Uzlopak** [comment]
I propose something like this 
https://github.com/fastify/fastify-swagger-ui/pull/20

File: test/hooks.test.js (hunk 1 above)
Comment (climba03003) [reply test/hooks.test.js:85]: `onSend` hook is more generic and not limited in changing `swagger` output only.
They can also mutate the `config`, etc.

`uiHook` like `onRequest` and `preHandler` also spread across all the route.
The users should be aware on the limitation and they need to check the `url` to match dedicated route.

File: test/hooks.test.js [hunk 2]
```diff
@@ -74,6 +74,35 @@ test('hooks on static swagger', async t => {
...
+    uiHooks: {
+      onSend: function (request, reply, payload, done) {
+        const swagger = JSON.parse(payload)
+        swagger.description = 'TestOnSend'
+        const newPayload = JSON.stringify(swagger)
+        done(null, newPayload)
```
Comment (mcollina) [inline test/hooks.test.js:91]: I would not recommend anybody to do this approach either. What about YAML?
I think the best solution is to do #20

Comment (climba03003) [reply test/hooks.test.js:91]: Somehow, `yaml` is never be used inside the `swagger-ui`.
I don't know if it is needed inside this library.

File: test/hooks.test.js (hunk 1 above)
Comment (Uzlopak) [reply test/hooks.test.js:85]: The unit test already shows me, that it is not intuitive to implement it correctly.

Comment (climba03003) [reply test/hooks.test.js:85]: I would argue that even with `onRequest` or `preHandler`.
People would make it wrongly.

**Comment by jonaskello** [comment]
Closing in favor of #20

**Pull request closed without merging**